compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BUnionType.java

```diff
@@ -217,9 +218,16 @@ public boolean isPureType() {
 
     private static LinkedHashSet<BType> toFlatTypeSet(LinkedHashSet<BType> types) {
         return types.stream()
-                .flatMap(type -> type.tag == TypeTags.UNION ?
+                .flatMap(type -> type.tag == TypeTags.UNION && !isTypeParamAvailable(type) ?
                         ((BUnionType) type).memberTypes.stream() : Stream.of(type))
                 .collect(Collectors.toCollection(LinkedHashSet::new));
     }
 
+    private static boolean isTypeParamAvailable(BType type) {
```

Comment:
Since we are checking if this particular type is a type param, shall we rename this method to `isTypeParam()`? Check the `isTypeParam()` method in `TypeParamAnalyzer`

Reply:
Changed